File: JavaScript/5_DataStructures/8_PriorityQueue.js
class Node {
  constructor(val, priority) {
    this.val = val;
    this.priority = priority;
  }
}

class PriorityQueue {
  constructor() {
    this.queue = [];
  }

  enqueue(val, priority) {
    const new_node = new Node(val, priority);
    this.queue.push(new_node);

    let currentIndex = this.queue.length - 1;
    let parentIndex;

    while (currentIndex > 0) {
      parentIndex = Math.floor((currentIndex - 1) / 2);
      if (this.queue[parentIndex].priority < this.queue[currentIndex].priority)
        return this;
      [this.queue[parentIndex], this.queue[currentIndex]] = [
        this.queue[currentIndex],
        this.queue[parentIndex],
      ];
      currentIndex = parentIndex;
    }
    return this;
  }

  dequeue() {
    let maxPriority = this.queue[0];
    let parent = this.queue.pop();
    let length = this.queue.length;
    if (length === 0) return maxPriority;
    let parentIndex = 0;
    this.queue[parentIndex] = parent;

    let left,
      leftIndex = 1,
      right,
      rightIndex = 2;
    let newIndex = parentIndex;
    while (leftIndex < length || rightIndex < length) {
      left = this.queue[leftIndex];
      right = this.queue[rightIndex];
      parent = this.queue[parentIndex];

      if (left && left.priority < parent.priority) {
        newIndex = leftIndex;
      }
      if (
        right &&
        right.priority < left.priority &&
        right.priority < parent.priority
      ) {
        newIndex = rightIndex;
      }

      if (newIndex === parentIndex) break;
      [this.queue[newIndex], this.queue[parentIndex]] = [
        this.queue[parentIndex],
        this.queue[newIndex],
      ];
      parentIndex = newIndex;

      leftIndex = 2 * parentIndex + 1;
      rightIndex = 2 * parentIndex + 2;
    }
    return maxPriority;
  }
}

let PQ = new PriorityQueue();

/*
Priority Queue is same like Binary Heap.

but its Minimum Binary Heap with respect priority => 1 = high priority, 5 = less priority

Time Complexity => O(1) getting
                => O(log n) enqueing 
*/

File: Python/5_DataStructures/8_PriorityQueue.py
class Node(object):
	def __init__(self, val, priority):
		self.val = val
		self.priority = priority

	def __str__(self):
		return f'{self.val}-{self.priority}'

	def __repr__(self):
		return f'{self.val}-{self.priority}'
		

class PriorityQueue(object):
	def __init__(self):
		self.queue = []

	def enqueue(self, val, priority):
		if len(self.queue) == 0:
			return self.queue.append(Node(val, priority)) != False

		self.queue.append(Node(val, priority))

		child_idx = len(self.queue) - 1

		while child_idx > 0:
			parent_idx = math.floor((child_idx-1)/2)
			child = self.queue[child_idx]
			parent = self.queue[parent_idx]

			if parent.priority <= child.priority: break

			self.queue[parent_idx], self.queue[child_idx] = self.queue[child_idx], self.queue[parent_idx]

			child_idx = parent_idx

		return True

	def dequque(self):
		if len(self.queue) == 0: return False

		new_root = self.queue.pop()
		if len(self.queue) == 0: return new_root
		max_value = self.queue[0]
		self.queue[0] = new_root
		length = len(self.queue)

		parent_idx = 0
		left_idx, right_idx = 1, 2
		new_idx = 0

		while True:
			parent = self.queue[parent_idx]

			if left_idx >= length: break
			left = self.queue[left_idx]
			if(left.priority < parent.priority):
				new_idx = left_idx

			if right_idx >= length: break
			right = self.queue[right_idx]
			if(right.priority < left.priority and right.priority < parent.priority):
				new_idx = right_idx

			if new_idx == parent_idx: break

			self.queue[new_idx], self.queue[parent_idx] = self.queue[parent_idx], self.queue[new_idx]

			parent_idx = new_idx
			left_idx = (2 * parent_idx) + 1
			right_idx = (2 * parent_idx) + 2

		return max_value

		